src/fromager/sdist.py

```diff
@@ -259,6 +278,150 @@ def handle_requirement(
     return resolved_version
 
 
+def _resolve_from_graph(
+    ctx: context.WorkContext,
+    prev_graph: DependencyGraph | None,
+    parent_req: Requirement | None,
+    req: Requirement,
+    req_type: RequirementType,
+    pre_built: bool,
+) -> tuple[str, Version] | None:
+    # we have already resolved top level reqs before bootstrapping
+    # so they should already be in the root node
+    if req_type == RequirementType.TOP_LEVEL:
+        for edge in ctx.dependency_graph.get_root_node().get_outgoing_edges(
+            req.name, RequirementType.TOP_LEVEL
+        ):
+            if edge.req == req:
+                return (
+                    edge.destination_node.download_url,
+                    edge.destination_node.version,
+                )
+        # this should never happen
```

Comment:
Maybe say why, because I had to sit for a minute to think about it. And maybe raise an exception in that case?